### PR TITLE
Floating Point Compare

### PR DESCRIPTION
**Pull request by rdaly525**

### CODE REVIEW

File: lassen/sim.py [hunk 1]
```diff
@@ -264,6 +302,7 @@ def __call__(self, inst: Inst, \
 
             # calculate 1-bit result
             res_p = cond(inst.cond, alu_res_p, lut_res, Z, N, C, V)
+            #print(alu_res,Z,N,res_p,inst.cond)
```
Comment (cdonovick) [inline lassen/sim.py]: remove

File: lassen/asm.py [hunk 2]
```diff
@@ -73,9 +73,34 @@ def fgetmant ():
 def fp_add(ra_mode=Mode.BYPASS, rb_mode=Mode.BYPASS):
     return inst(ALU.FP_add, ra_mode=ra_mode, rb_mode=rb_mode)
 
+def fp_sub(ra_mode=Mode.BYPASS, rb_mode=Mode.BYPASS):
+    return inst(ALU.FP_sub, ra_mode=ra_mode, rb_mode=rb_mode)
+
 def fp_mult(ra_mode=Mode.BYPASS, rb_mode=Mode.BYPASS):
     return inst(ALU.FP_mult, ra_mode=ra_mode, rb_mode=rb_mode)
 
+def fp_cmp(cond):
```
Comment (cdonovick) [inline lassen/asm.py]: Why not have a dedicate fp_cmp instruction so we don't have to set the Z strangely on sub

File: lassen/sim.py [hunk 3]
```diff
@@ -206,8 +233,15 @@ def alu(inst:Inst, a:Data, b:Data, d:Bit) -> (Data, Bit, Bit, Bit, Bit,
...
 
-        Z = res == 0
         N = Bit(res[-1])
+        if (alu == ALU.FP_sub) | (alu == ALU.FP_add) | (alu == ALU.FP_mult):
+            Z = fp_is_zero(res)
+        else:
+            Z = (res == 0)
+
+        #Nicely handles infinities for comparisons
+        if (alu == ALU.FP_sub) & (a_inf & b_inf) & (a_neg == b_neg):
```
Comment (cdonovick) [inline lassen/sim.py]: Would rather have if 
```
if (alu == ALU.FP_cmp) & ...
```
Than break the Z flag

File: lassen/asm.py (hunk 2 above)
Comment (rdaly525) [reply lassen/asm.py]: Agreed. fixed

**Comment by rdaly525** [comment]
@leonardt, Can you take a look here? I am not sure why there is a width mismatch. I added a few more condition codes to the ISA which causes the number of required bits (for Cond) to go from 4->5. the assembler produces a layout which correctly indicates the number 5, but magma is claiming a width mismatch.

**Comment by leonardt** [comment]
@rdaly525 it looks like the in the generated magma code, `assemble` isn't being called on the enumerated condition codes, I will investigate why this is the case.

**Comment by rdaly525** [comment]
Caleb said that you can recreate this by simply adding a new field in the Cond enum

**Comment by rdaly525** [comment]
Depends on https://github.com/phanrahan/peak/pull/54

**Comment by rdaly525** [comment]
@cdonovick, can you re-review?

**Comment by rdaly525** [comment]
@leonardt, should this branch be failing without https://github.com/phanrahan/peak/pull/54?

**Comment by rdaly525** [comment]
@Kuree, how do I update the kite build? This will need the following branch to work: https://github.com/phanrahan/peak/pull/54

**Comment by Kuree** [comment]
@rdaly525 It should be the same way as you update the travis. You can modify this line to use your own branch: https://github.com/StanfordAHA/lassen/blob/master/requirements.txt#L2

**Comment by rdaly525** [comment]
@Kuree/ @leonardt I am seeing that RTL is reporting an error for some of the fp_binary tests in test_pe while utilizing the FPadd/mul IP. Could you debug?

**Comment by Kuree** [comment]
@rdaly525 Do you have access to `kiwi`? I can help you setup there so you can debug locally.

From the log it seems like `FP_sub` is failing. Since this PR adds the `sub` instruction, it's likely that there is an inconsistency between the functional model and RTL.

**Comment by rdaly525** [comment]
I do not have access to Kiwi I think... Could you make me an account?

**Comment by Kuree** [comment]
@rdaly525 I've created an account for you on kiwi. You should be able to use your Stanford credential to log in. I highly recommend you to use `/nobackupkiwi/rdaly525` instead of your home directory for big files.

**Comment by rdaly525** [comment]
Caleb was able to track down that the FP subtract was actually doing an integer subtract. I suspect this is due to magma.BFloat does not have an implementation for subtract and is utilizing bitvector subtract

**Comment by leonardt** [comment]
@rdaly525 correct, sub has not been added to magma bfloat (it only support add/mul as per original goals), is there a corresponding coreir fp sub primitive in coreir that I can use to test it an implementation?

File: tests/test_pe.py [hunk 4]
```diff
@@ -282,9 +285,30 @@ def test_fp_binary_op(op,args):
...
+    op('le',  lambda x, y: x <= y),
+    op('eq',  lambda x, y: x == y),
+    op('ne',  lambda x, y: x != y),
+])
+def test_fp_cmp(xy,op):
+    inst = getattr(asm,f"fp_{op.inst}")()
+    in0,in1 = xy
+    out = op.func(in0,in1)
+    _, res_p, _ = pe(inst, BFloat16.reinterpret_as_bv(in0), BFloat16.reinterpret_as_bv(in1))
+    assert res_p == out
```
Comment (Kuree) [inline tests/test_pe.py]: Can you add RTL tests as well?

**Comment by rdaly525** [comment]
@Kuree, could you approve this request? I have another branch which adds in the rtl tests. That branch is failing due to the precision issue, but I think we can resolve that in a separate PR

**Pull request merged**